pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -705,26 +705,25 @@ func (cb *ClusterBuilder) buildDefaultPassthroughCluster() *cluster.Cluster {
 			v3.HttpProtocolOptionsType: passthroughHttpProtocolOptions,
 		},
 	}
-	cb.applyConnectionPool(cb.req.Push.Mesh, NewMutableCluster(cluster), &networking.ConnectionPoolSettings{})
+	cb.applyConnectionPool(cb.req.Push.Mesh, newClusterWrapper(cluster), &networking.ConnectionPoolSettings{}, nil)
 	cb.applyMetadataExchange(cluster)
 	return cluster
 }
 
-// applyH2Upgrade function will upgrade outbound cluster to http2 if specified by configuration.
-func (cb *ClusterBuilder) applyH2Upgrade(opts buildClusterOpts, connectionPool *networking.ConnectionPoolSettings) {
-	if cb.shouldH2Upgrade(opts.mutable.cluster.Name, opts.direction, opts.port, opts.mesh, connectionPool) {
-		cb.setH2Options(opts.mutable)
+// applyH2Upgrade function will upgrade cluster to http2 if specified by configuration.
+func (cb *ClusterBuilder) applyH2Upgrade(mc *clusterWrapper, port *model.Port,
+	mesh *meshconfig.MeshConfig, connectionPool *networking.ConnectionPoolSettings) bool {
+	if cb.shouldH2Upgrade(mc.cluster.Name, port, mesh, connectionPool) {
+		cb.setH2Options(mc)
+		return true
 	}
+	return false
 }
 
 // shouldH2Upgrade function returns true if the cluster should be upgraded to http2.
-func (cb *ClusterBuilder) shouldH2Upgrade(clusterName string, direction model.TrafficDirection, port *model.Port, mesh *meshconfig.MeshConfig,
+func (cb *ClusterBuilder) shouldH2Upgrade(clusterName string, port *model.Port, mesh *meshconfig.MeshConfig,
 	connectionPool *networking.ConnectionPoolSettings,
 ) bool {
-	if direction != model.TrafficDirectionOutbound {
-		return false
```

Comment:
this maybe a breaking change for inbound cluster. 

But we respect useClientProtocol for inbound, so I think same should apply to h2UpgradePolicy

Reply:
useClientProtocol is different from upgrade. useClientProtocol is used in cases where server supports both H1/H2 and the connection from Envoy -> server should be chosen based on client protocol. 
H2 upgrade is used to upgrade connections from client Envoy -> Server Envoy to get the benefits of in that hop.
This will break if server does not support H2 but still user wants to optimize the path of client Envoy -> server Envoy.

Reply:
Detailed explanation, can we add some more comments to the API

Reply:
Why was this merged? I agree with Rama's comment "This will break if server does not support H2 but still user wants to optimize the path of client Envoy -> server Envoy.".

This is a breaking change as far as I can tell? It doesn't seem like this comment was addressed?

Reply:
hmm. I thought this comment was addressed. @hzxuzhonghu can you PTAL?

Reply:
ack, thanks Rama